README.md
# Event Data Status Service

Collect heartbeats from across Event Data.

## Docker

A Docker image is used for deployment. A Docker Compose file is included for testing only.

## Tinkering

Start the server (NB you must specify port mappings for `docker compose-run`)

    docker-compose run -w /code --publish "8003:8003" test lein run

Then you can post updates:

    curl --verbose --header "Content-Type: text/plain" --data "5" -H "Authorization: Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyIxIjoiMSIsInN1YiI6Indpa2lwZWRpYSJ9.w7zV2vtKNzrNDfgr9dfRpv6XYnspILRli_V5vd1J29Q" http://127.0.0.1:8003/status/my-service/my-cmponent/my-facet

## Testing

Unit tests:

  - `time docker-compose run -w /code test lein test :unit`

Component tests:

  - `time docker-compose run -w /code test lein test :component`

## Configuration

The following configuration keys must be set in any code that uses these libraries:

| Environment variable | Description                         | Default |
|----------------------|-------------------------------------|---------|
| `REDIS_HOST`         | Redis host                          |         |
| `REDIS_PORT`         | Redis port                          |         |
| `REDIS_DB`           | Redis DB number                     | 0       |
| `S3_KEY`             | AWS Key Id                          |         |
| `S3_SECRET`          | AWS Secret Key                      |         |
| `S3_BUCKET_NAME`     | AWS S3 bucket name                  |         |
| `S3_REGION_NAME`     | AWS S3 bucket region name           |         |
| `JWT_SECRETS`        | Comma-separated list of JTW Secrets |         |

## License

Copyright © Crossref

Distributed under the The MIT License (MIT).


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-12-12
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-12-12
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/event-data-status/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/event-data-status/compare/0.1.0...0.1.1


test/event_data_status/core_test.clj
(ns event-data-status.core-test
  (:require [clojure.test :refer :all]
            [event-data-status.core :refer :all]))



test/event_data_status/server_tests.clj
(ns event-data-status.server-tests
  "Tests for the server namespace.
   Most component tests work at the Ring level, including routing.
   These run through all the middleware including JWT extraction."
  (:require [clojure.test :refer :all]
            [clojure.data.json :as json]
            [clojure.tools.logging :as l]
            [event-data-common.storage.store :as store]
            [event-data-common.jwt :as jwt]
            [event-data-status.server :as server]
            [ring.mock.request :as mock]
            [clj-time.core :as clj-time]
            [clj-time.format :as clj-time-format]
            [clojure.java.io :as io])
  (:import [com.auth0.jwt JWTSigner JWTVerifier]))

(defn clear-storage-fixture
  [f]
  ; Clear all keys from Redis.
  (doseq [k (store/keys-matching-prefix @server/redis-store "")]
    (store/delete @server/redis-store k))
  (f))

(use-fixtures :each clear-storage-fixture)

(def jwt-signer
  "Build a signer that corresponds to the secret set in the config via docker-compose.yml ."
  (delay (jwt/build "TEST")))

(deftest ^:component increment-errors
  (testing "Malformed input produces a 400"
    (let [auth (str "Bearer " (jwt/sign @jwt-signer {}))
          empty-result (-> (mock/request :post "/status/my-service/my-component/my-facet")
                           (mock/header "authorization" auth)
                           (mock/body "")
                           (@server/app)
                           :status)

          not-number-result (-> (mock/request :post "/status/my-service/my-component/my-facet")
                             (mock/header "authorization" auth)
                             (mock/body "Five")
                             (@server/app)
                             :status)

          unauthenticated-number (-> (mock/request :post "/status/my-service/my-component/my-facet")
                                   (mock/header "authorization" "Grizzly Bearer")
                                   (mock/body "5")
                                   (@server/app)
                                   :status)]

          (is (= 400 empty-result) "Empty POST should be malfomed")
          (is (= 400 not-number-result) "Non-number POST should be allowed")
          (is (= 401 unauthenticated-number) "Correct number with no auth should be 403"))))

(deftest ^:component roundtrip-increment
  (testing "Counts can be incremented"
    ; Freeze time.
      (let [auth (str "Bearer " (jwt/sign @jwt-signer {}))]
        (clj-time/do-at (clj-time/date-time 1986 05 02 12 00)
          (let [; First get zero value.

                pre-result (->
                             (mock/request :get "/status/my-service/my-component/my-facet")
                             (mock/header "authorization" auth)
                             (@server/app)
                             :body)

                ; Then increment by 5.
                inc-result-1 (->
                               (mock/request :post "/status/my-service/my-component/my-facet")
                               (mock/header "authorization" auth)
                               (mock/body "5")
                               (@server/app)
                               :body)

                ; Then get new value.
                result-2 (->
                           (mock/request :get "/status/my-service/my-component/my-facet")
                           (mock/header "authorization" auth)
                           (@server/app)
                           :body)

                ; Then increment by 6.
                inc-result-3 (->
                               (mock/request :post "/status/my-service/my-component/my-facet")
                               (mock/header "authorization" auth)
                               (mock/body "6")
                               (@server/app)
                               :body)

                ; Then get new value.
                result-4 (->
                           (mock/request :get "/status/my-service/my-component/my-facet")
                           (mock/header "authorization" auth)
                           (@server/app)
                           :body)]

            (is (= "0" pre-result) "GET Status for service/component/facet should be zero initially.")
            (is (= "5" inc-result-1) "Increment should return new value.")
            (is (= "5" result-2) "GET Status should reflect new value.")
            (is (= "11" inc-result-3) "Increment again should return new value.")
            (is (= "11" result-4) "GET Status should again reflect new value.")))

        ; One minute later.
        (clj-time/do-at (clj-time/date-time 1986 05 02 12 01)
          (let [pre-result (->
                             (mock/request :get "/status/my-service/my-component/my-facet")
                             (mock/header "authorization" auth)
                             (@server/app)
                             :body)]

            (is (= "0" pre-result) "GET Status should be reset the next minute."))))))

(deftest ^:component events-default-24h    
    ; Set one key in this date range so it exists.  
    (let [auth (str "Bearer " (jwt/sign @jwt-signer {}))]
      (clj-time/do-at (clj-time/date-time 1986 05 02 12 01)
        ; Same thing twice in the same minute should accumulate.
        (-> (mock/request :post "/status/my-service/my-component/my-facet")
          (mock/header "authorization" auth)
          (mock/body "1")
          (@server/app))

        (-> (mock/request :post "/status/my-service/my-component/my-facet")
          (mock/header "authorization" auth)
          (mock/body "2")
          (@server/app)))

      ; Update that and another one the next minute.
      (clj-time/do-at (clj-time/date-time 1986 05 02 12 02)
        (-> (mock/request :post "/status/my-service/my-component/my-facet")
          (mock/header "authorization" auth)
          (mock/body "10")
          (@server/app))

        (-> (mock/request :post "/status/my-service/my-component/other-facet")
          (mock/header "authorization" auth)
          (mock/body "18")
          (@server/app))))

    (testing "Fetching events without arguments returns last 24 hours from now."
      ; Query one hour later.
      (clj-time/do-at (clj-time/date-time 1986 05 02 13 00)
        (let [result (->
                             (mock/request :get "/status")
                             (@server/app)
                             :body
                             json/read-str)]

          (is (= "1986-05-02T13:00:00.000Z" (result "end")) "Default end-date should be now")
          (is (= "1986-05-01T13:00:00.000Z" (result "start")) "Default start-date should be now minus 24 hours")

          (is (= (count (get-in result ["services" "my-service" "my-component" "my-facet"])) (* 60 24)) "Correct number of minutes should be returned")
          (is (= (count (get-in result ["services" "my-service" "my-component" "other-facet"])) (* 60 24)) "Correct number of minutes should be returned")

          ; Check that we got the right distinct values.
          (is (= #{0 3 10} (set (get-in result ["services" "my-service" "my-component" "my-facet"]))) "Distinct values should have been summed")
          (is (= #{0 18} (set (get-in result ["services" "my-service" "my-component" "other-facet"]))) "Distinct values should have been summed"))))

    (testing "Fetching custom hours returns that range."
      ; Query one hour later.
      (clj-time/do-at (clj-time/date-time 1986 05 02 12 03)
        (let [result (->
                             (mock/request :get "/status?hours=1")
                             (@server/app)
                             :body
                             json/read-str)]

          (is (= "1986-05-02T12:03:00.000Z" (result "end")) "Default end-date should be now")
          (is (= "1986-05-02T11:03:00.000Z" (result "start")) "Default start-date should be now minus 24 hours")

          (is (= (count (get-in result ["services" "my-service" "my-component" "my-facet"])) 60) "60 minutes should be returned for 1 hour")
          (is (= (count (get-in result ["services" "my-service" "my-component" "other-facet"])) 60) "60 minutes should be returned for 1 hour")

          ; Check that we got the right distinct values.
          (is (= #{0 3 10} (set (get-in result ["services" "my-service" "my-component" "my-facet"]))) "Distinct values should have been summed")
          (is (= #{0 18} (set (get-in result ["services" "my-service" "my-component" "other-facet"]))) "Distinct values should have been summed")

          (is (= [0 0 0 3 10] (take-last 5 (get-in result ["services" "my-service" "my-component" "my-facet"]))) "Correct values should be found at the most recent end of the result.")
          (is (= [0 0 0 0 18] (take-last 5 (get-in result ["services" "my-service" "my-component" "other-facet"]))) "Correct values should be found at the most recent end of the result."))))

    (testing "Fetching more than 24 hours is an error."
      (let [result-status-code (->
                           (mock/request :get "/status?hours=26")
                           (@server/app)
                           :status)]
        (is (= result-status-code 400) "More than 24 hours should be an error"))))



project.clj
(defproject event-data-status "0.1.1"
  :description "Crossref Event Data Status Service"
  :url "http://eventdata.crossref.org"
  :license {:name "The MIT License (MIT)"
            :url "https://opensource.org/licenses/MIT"}
  :dependencies [[org.clojure/clojure "1.7.0"]
                 [event-data-common "0.1.5"]
                 [org.clojure/data.json "0.2.6"]
                 [crossref-util "0.1.10"]
                 [http-kit "2.1.18"]
                 [http-kit.fake "0.2.1"]
                 [liberator "0.14.1"]
                 [compojure "1.5.1"]
                 [ring "1.5.0"]
                 [ring/ring-jetty-adapter "1.5.0"]
                 [ring/ring-servlet "1.5.0"]
                 [ring/ring-mock "0.3.0"]
                 [org.eclipse.jetty/jetty-server "9.4.0.M0"]
                 [overtone/at-at "1.2.0"]
                 [robert/bruce "0.8.0"]
                 [yogthos/config "0.8"]
                 [crossref/heartbeat "0.1.2"]
                 [com.auth0/java-jwt "2.2.1"]
                 [clj-time "0.12.2"]
                 [redis.clients/jedis "2.8.0"]
                 [metosin/scjsv "0.4.0"]
                 [org.clojure/tools.logging "0.3.1"]
                 [org.apache.logging.log4j/log4j-core "2.6.2"]
                 [org.slf4j/slf4j-simple "1.7.21"]
                 [org.clojure/core.async "0.2.395"]]
  :main ^:skip-aot event-data-status.core
  :target-path "target/%s"
  :test-selectors {:default (constantly true)
                   :unit :unit
                   :component :component
                   :integration :integration
                   :all (constantly true)}
  :jvm-opts ["-Duser.timezone=UTC"]
  :profiles {:uberjar {:aot :all}
             :prod {:resource-paths ["config/prod"]}
             :dev  {:resource-paths ["config/dev"]}})


src/event_data_status/server.clj
(ns event-data-status.server
  (:require [clojure.data.json :as json]
            [clojure.tools.logging :as l])
  (:require [org.httpkit.server :as server]
            [config.core :refer [env]]
            [compojure.core :refer [defroutes GET POST]]
            [ring.util.response :as ring-response]
            [ring.middleware.params :as middleware-params]
            [ring.middleware.content-type :as middleware-content-type]
            [ring.middleware.resource :as middleware-resource]
            [liberator.core :refer [defresource]]
            [liberator.representation :as representation]
            [clj-time.core :as clj-time]
            [clj-time.format :as clj-time-format]
            [clj-time.coerce :as clj-time-coerce]
            [clj-time.periodic :as clj-time-periodic]
            [clojure.java.io :refer [reader input-stream]]
            [event-data-common.jwt :as jwt]
            [event-data-common.storage.redis :as redis]
            [event-data-common.storage.s3 :as s3]
            [event-data-common.storage.store :as store]
            [clojure.core.async :as async]
            [overtone.at-at :as at-at])
  (:import
           [java.net URL MalformedURLException InetAddress])
  (:gen-class))

(def redis-prefix
  "Unique prefix applied to every key."
  "status:")

(def date-formatter (clj-time-format/formatters :date-hour-minute))

; (def minute-formatter (clj-time-format/formatter "yyyyMMddHHmm"))
(def minute-formatter-length 16)

(def default-redis-db-str "1")

(defn key-for
  "Generate Redis key for the service-component-facet now."
  [^String service ^String component ^String facet]
  (let [now (clj-time/now)
        now-str (clj-time-format/unparse date-formatter now)]
    (str now-str "/" service "/" component "/" facet)))

(def redis-store
  "A redis connection for storing subscription and short-term information."
  (delay (redis/build redis-prefix (:redis-host env) (Integer/parseInt (:redis-port env)) (Integer/parseInt (get env :redis-db default-redis-db-str)))))

; Heartbeat
(def schedule-pool (at-at/mk-pool))

; Websocket things
(def channel-hub (atom {}))
(def pubsub-channel-name "__status__broadcast")

(defn broadcast
  "Send event to all websocket listeners."
  [data]
    (doseq [[channel channel-options] @channel-hub]
      (server/send! channel data)))

; Get current count.
(defresource get-status-coordinate
  [service component facet]
  :allowed-methods [:get]
  :available-media-types ["application/json"]
  :handle-ok (fn [context]
              (let [k (key-for service component facet)
                    v (store/get-string @redis-store k)
                    v-str (str (or v 0))]
                v-str)))

(defn ping!
  "Enact the update and return the new value."
  [service component facet value]
  (try
    (let [k (key-for service component facet)
          new-value (redis/incr-key-by!? @redis-store k value)]
      ; Broadcast to websockets via pubsub too.
      (redis/publish-pubsub @redis-store pubsub-channel-name (str service "/" component "/" facet ";" value))
      new-value)
    (catch Exception e (l/error "Failed to ping" service component facet value "because" (.getMessage e)))))

; Increment current count.
(defresource post-status-coordinate
  [service component facet]
  :allowed-methods [:post]
  :available-media-types ["application/json"]
  :authorized? (fn [ctx]
                 ; Any validated claim OK to post.
                 (-> ctx :request :jwt-claims))
  :malformed? (fn [ctx]
               (try
                 (let [value (Integer/parseInt (slurp (get-in ctx [:request :body])))]
                   [false {::value value}])
                  (catch NumberFormatException _ true)))
  :post! (fn [ctx]
           (let [new-value (ping! service component facet (::value ctx))]
             {::new-value (str new-value)}))
  :handle-created (fn [ctx]
              (::new-value ctx)))

(defn build-minute-structure
  "For a service/component/facet structure and a seq of minute strings, return a seq of values, filling in missing zeroes."
  [[service component facet] minute-prefixes]
  (let [k (str service "/" component "/" facet)]
    (map #(Integer/parseInt (or (store/get-string @redis-store (str % "/" k)) "0")) minute-prefixes)))

(defn build-structure
  "Take a seq of clj-time minutes and return a tree of {service {component {facet {date-time count}}}}"
  [start-date num-minutes]
  (let [minutes (take num-minutes (clj-time-periodic/periodic-seq start-date (clj-time/minutes 1)))
        
        ; as seq of prefixes in Redis.
        ; prefixes are well-formed dates.
        minute-prefixes (map #(clj-time-format/unparse date-formatter %) minutes)

        ; map of {minute matching-keys}
        keys-per-minute (into (sorted-map) (map #(vector % (store/keys-matching-prefix @redis-store %)) minute-prefixes))

        ; get the set of keys present
        all-keys (distinct (mapcat second keys-per-minute))

        ; all keys as triples of [service component facet] ignoring the datetime
        service-keys (map #(rest (.split % "/")) all-keys)

        ; into {service {component {facet minute-structure}}}
        ; where minute-structure is a seq of counts for *all* minutes in date range.
        tree (reduce (fn [acc service-component-facet]
          (assoc-in
            acc
            service-component-facet
            (build-minute-structure service-component-facet minute-prefixes))) {} service-keys)]
    tree))

(defresource get-status-range
  []
  :allowed-methods [:get]
  :available-media-types ["application/json"]
  :malformed? (fn [ctx]
                (let [supplied-end (get-in ctx [:request :params "end"])
                      supplied-hours (get-in ctx [:request :params "hours"])

                      end (if supplied-end
                                (clj-time-format/parse date-formatter supplied-end)
                                (clj-time/now))

                      hours (if supplied-hours
                              (Integer/parseInt supplied-hours)
                              24)

                      start (clj-time/minus end (clj-time/hours hours))

                      ok (<= hours 24)]
                  [(not ok) {::start start ::end end ::hours hours}]))

  :handle-ok (fn [ctx]
              (let [num-minutes (* 60 (::hours ctx))
                    structure (build-structure (::start ctx) num-minutes)]
              {:start (str (::start ctx))
               :end (str (::end ctx))
               :services structure})))


(defn socket-handler [request]
  (l/info "Socket handler")
  (server/with-channel request channel
    (server/on-close channel (fn [status]
                               (swap! channel-hub dissoc channel)))

    (server/on-receive channel (fn [data]
                                 ; Any input is a subscribe.
                                  (swap! channel-hub assoc channel {})))))

(defroutes app-routes
  (GET "/socket" [] socket-handler)
  (GET "/status" [] (get-status-range))
  (POST "/status/:service/:component/:facet" [service component facet] (post-status-coordinate service component facet))
  (GET "/status/:service/:component/:facet" [service component facet] (get-status-coordinate service component facet))
  (GET "/" [] (ring-response/redirect "http://eventdata.crossref.org")))


(defn wrap-cors
  "Middleware to add a liberal CORS header."
  [handler]
  (fn [request]
    (let [response (handler request)]
      (when response
        (assoc-in response [:headers "Access-Control-Allow-Origin"] "*")))))

(def app
  ; Delay construction to runtime for secrets config value.
  (delay
    (-> app-routes
       middleware-params/wrap-params
       (middleware-resource/wrap-resource "public")
       (jwt/wrap-jwt (:jwt-secrets env))
       (middleware-content-type/wrap-content-type)
       (wrap-cors))))

(defn run-server []
  (let [port (Integer/parseInt (:port env))]
    (l/info "Start pusub listener.")

    ; Listen on pubsub and send to all listening websockets.
    (async/thread (redis/subscribe-pubsub @redis-store pubsub-channel-name #(broadcast %)))

    ; Ping the heartbeat every 10 seconds.
    ; (Every instance in the cluster will do this)
    (at-at/every 10000 #(ping! "status" "heartbeat" "ping" 1) schedule-pool)

    (l/info "Start server on " port)
    (server/run-server @app {:port port})))
